File: packages/mcp-servers/waveform-mcp/src/waveform_mcp/parser.py
"""VCD/FST waveform parser with timescale normalization.

Uses vcdvcd for VCD parsing. FST is handled via GTKWave's fst2vcd conversion
when available, falling back to a "not supported" message.
"""

from __future__ import annotations

import re
import shutil
import subprocess
import tempfile
from dataclasses import dataclass, field
from pathlib import Path

# ── Timescale utilities ───────────────────────────────────────────────────────

_TS_UNITS = {"fs": 1e-6, "ps": 1e-3, "ns": 1.0, "us": 1e3, "ms": 1e6, "s": 1e9}


def timescale_to_ns(ts: str) -> float:
    """Parse a VCD timescale string and return the ns multiplier.

    e.g. "1ns" → 1.0,  "10ps" → 0.01,  "1us" → 1000.0
    """
    ts = ts.strip().lower()
    for unit, factor in _TS_UNITS.items():
        if ts.endswith(unit):
            try:
                value = float(ts[: -len(unit)].strip() or "1")
                return value * factor
            except ValueError:
                return factor
    return 1.0  # default: assume ns


# ── Signal data model ─────────────────────────────────────────────────────────

@dataclass
class SignalTrace:
    name: str
    width: int                              # bit width
    scope: str                              # HDL scope path
    tv: list[tuple[float, str]] = field(default_factory=list)   # (time_ns, value_str)

    @property
    def is_clock(self) -> bool:
        """Heuristic: regular binary toggle with ≥20 transitions."""
        if self.width != 1 or len(self.tv) < 20:
            return False
        vals = [v for _, v in self.tv[:40]]
        # Must alternate 0/1
        alternating = all(
            vals[i] != vals[i + 1] for i in range(min(10, len(vals) - 1))
        )
        if not alternating:
            return False
        # Period must be consistent (std < 5% of mean)
        periods = [self.tv[i + 1][0] - self.tv[i][0] for i in range(min(20, len(self.tv) - 1))]
        if not periods:
            return False
        mean = sum(periods) / len(periods)
        if mean <= 0:
            return False
        std = (sum((p - mean) ** 2 for p in periods) / len(periods)) ** 0.5
        return (std / mean) < 0.05

    @property
    def clock_period_ns(self) -> float:
        """Full clock period in ns (2 × half-period)."""
        if len(self.tv) < 2:
            return 0.0
        half_periods = [
            self.tv[i + 1][0] - self.tv[i][0]
            for i in range(min(10, len(self.tv) - 1))
        ]
        return 2.0 * (sum(half_periods) / len(half_periods)) if half_periods else 0.0


@dataclass
class WaveformMetadata:
    format: str                             # "vcd" | "fst"
    file_path: str
    duration_ns: float
    ts_factor_ns: float                     # timescale factor in ns
    signals: dict[str, SignalTrace] = field(default_factory=dict)

    @property
    def clocks(self) -> list[dict]:
        result = []
        for name, sig in self.signals.items():
            if sig.is_clock:
                period = sig.clock_period_ns
                result.append({
                    "name":      name,
                    "period_ns": round(period, 3),
                    "freq_mhz":  round(1000.0 / period, 3) if period > 0 else 0,
                })
        return result

    def to_summary(self) -> dict:
        return {
            "format":       self.format,
            "duration_ns":  round(self.duration_ns, 3),
            "signal_count": len(self.signals),
            "clocks":       self.clocks,
            "signals": [
                {
                    "name":  name,
                    "width": sig.width,
                    "scope": sig.scope,
                }
                for name, sig in list(self.signals.items())[:200]   # cap for large projects
            ],
        }


# ── VCD parser ────────────────────────────────────────────────────────────────

def parse_vcd(file_path: str) -> WaveformMetadata:
    """Parse a VCD file and return WaveformMetadata with all signal traces.

    Uses vcdvcd library for robust VCD parsing.
    """
    try:
        from vcdvcd import VCDVCD
    except ImportError as e:
        raise RuntimeError("vcdvcd not installed: pip install vcdvcd") from e

    vcd = VCDVCD(file_path, only_sigs=False, store_tvs=True)

    ts_str = getattr(vcd, "timescale", "1ns")
    if isinstance(ts_str, dict):
        # Some versions return {"timescale": "1ns"}
        ts_str = ts_str.get("timescale", "1ns")
    ts_factor = timescale_to_ns(str(ts_str))

    endtime_raw = getattr(vcd, "endtime", 0) or 0
    duration_ns = float(endtime_raw) * ts_factor

    signals: dict[str, SignalTrace] = {}
    for sig_name in vcd.signals:
        try:
            data = vcd[sig_name]
            scope_parts = sig_name.rsplit(".", 1)
            scope = scope_parts[0] if len(scope_parts) > 1 else ""
            tv_ns = [(float(t) * ts_factor, v) for t, v in (data.tv or [])]
            signals[sig_name] = SignalTrace(
                name=sig_name,
                width=int(getattr(data, "size", 1)),
                scope=scope,
                tv=tv_ns,
            )
        except Exception:
            continue  # skip signals that fail to parse

    return WaveformMetadata(
        format="vcd",
        file_path=file_path,
        duration_ns=duration_ns,
        ts_factor_ns=ts_factor,
        signals=signals,
    )


def parse_fst(file_path: str) -> WaveformMetadata:
    """Parse an FST file by converting to VCD via GTKWave's fst2vcd tool."""
    if not shutil.which("fst2vcd"):
        raise RuntimeError(
            "fst2vcd not found. Install GTKWave (brew install gtkwave) to enable FST parsing."
        )
    with tempfile.NamedTemporaryFile(suffix=".vcd", delete=False) as tmp:
        tmp_path = tmp.name

    subprocess.run(
        ["fst2vcd", "-o", tmp_path, file_path],
        check=True,
        capture_output=True,
    )
    meta = parse_vcd(tmp_path)
    meta.format = "fst"
    meta.file_path = file_path
    Path(tmp_path).unlink(missing_ok=True)
    return meta


def parse_waveform(file_path: str) -> WaveformMetadata:
    """Auto-detect format and parse waveform file."""
    suffix = Path(file_path).suffix.lower()
    if suffix == ".vcd":
        return parse_vcd(file_path)
    if suffix == ".fst":
        return parse_fst(file_path)
    raise ValueError(f"Unsupported waveform format: {suffix}. Supported: .vcd, .fst")

File: packages/mcp-servers/waveform-mcp-rs/src/parser.rs
//! VCD/FST waveform parser with timescale normalization.
//!
//! VCD is parsed natively using the `vcd` crate.
//! FST is handled via GTKWave's `fst2vcd` conversion tool.

use std::collections::HashMap;
use std::fs::File;
use std::io::BufReader;

use anyhow::{bail, Context, Result};
use serde::Serialize;
use vcd::{Command, IdCode, Parser, ScopeItem, TimescaleUnit, Value};

// ── Signal data model ────────────────────────────────────────────────────────

#[derive(Clone, Debug, Serialize)]
pub struct SignalTrace {
    pub name: String,
    pub width: u32,
    pub scope: String,
    /// (time_ns, value_str) pairs sorted by time
    pub tv: Vec<(f64, String)>,
}

impl SignalTrace {
    /// Heuristic: regular binary toggle with ≥20 transitions and consistent period.
    pub fn is_clock(&self) -> bool {
        if self.width != 1 || self.tv.len() < 20 {
            return false;
        }
        let vals: Vec<&str> = self.tv.iter().take(40).map(|(_, v)| v.as_str()).collect();
        if vals.len() < 11 {
            return false;
        }
        // Must alternate between 0 and 1
        let alternating = vals.windows(2).take(10).all(|w| w[0] != w[1]);
        if !alternating {
            return false;
        }
        // Period must be consistent (std < 5% of mean)
        let periods: Vec<f64> = self
            .tv
            .windows(2)
            .take(20)
            .map(|w| w[1].0 - w[0].0)
            .collect();
        if periods.is_empty() {
            return false;
        }
        let mean = periods.iter().sum::<f64>() / periods.len() as f64;
        if mean <= 0.0 {
            return false;
        }
        let variance =
            periods.iter().map(|&p| (p - mean).powi(2)).sum::<f64>() / periods.len() as f64;
        (variance.sqrt() / mean) < 0.05
    }

    /// Full clock period in ns (2 × average half-period).
    pub fn clock_period_ns(&self) -> f64 {
        if self.tv.len() < 2 {
            return 0.0;
        }
        let half_periods: Vec<f64> = self
            .tv
            .windows(2)
            .take(10)
            .map(|w| w[1].0 - w[0].0)
            .collect();
        if half_periods.is_empty() {
            return 0.0;
        }
        2.0 * half_periods.iter().sum::<f64>() / half_periods.len() as f64
    }
}

// ── Waveform metadata ────────────────────────────────────────────────────────

#[derive(Clone, Debug)]
pub struct WaveformMeta {
    pub format: String,
    pub file_path: String,
    pub duration_ns: f64,
    pub signals: HashMap<String, SignalTrace>,
}

impl WaveformMeta {
    pub fn clocks(&self) -> Vec<serde_json::Value> {
        let mut clocks: Vec<_> = self
            .signals
            .values()
            .filter(|s| s.is_clock())
            .map(|s| {
                let period = s.clock_period_ns();
                let freq_mhz = if period > 0.0 { 1000.0 / period } else { 0.0 };
                serde_json::json!({
                    "name": s.name,
                    "period_ns": (period * 1000.0).round() / 1000.0,
                    "freq_mhz":  (freq_mhz * 1000.0).round() / 1000.0,
                })
            })
            .collect();
        clocks.sort_by(|a, b| {
            a["name"]
                .as_str()
                .unwrap_or("")
                .cmp(b["name"].as_str().unwrap_or(""))
        });
        clocks
    }

    pub fn to_summary(&self) -> serde_json::Value {
        let mut names: Vec<&str> = self.signals.keys().map(|s| s.as_str()).collect();
        names.sort_unstable();

        let signals: Vec<_> = names
            .iter()
            .take(200)
            .map(|&n| {
                let s = &self.signals[n];
                serde_json::json!({
                    "name":  s.name,
                    "width": s.width,
                    "scope": s.scope,
                })
            })
            .collect();

        serde_json::json!({
            "format":       self.format,
            "duration_ns":  (self.duration_ns * 1000.0).round() / 1000.0,
            "signal_count": self.signals.len(),
            "clocks":       self.clocks(),
            "signals":      signals,
        })
    }
}

// ── Helpers ──────────────────────────────────────────────────────────────────

fn timescale_to_ns(magnitude: u32, unit: TimescaleUnit) -> f64 {
    let factor = match unit {
        TimescaleUnit::FS => 1e-6,
        TimescaleUnit::PS => 1e-3,
        TimescaleUnit::NS => 1.0,
        TimescaleUnit::US => 1e3,
        TimescaleUnit::MS => 1e6,
        TimescaleUnit::S => 1e9,
    };
    magnitude as f64 * factor
}

fn value_to_char(v: &Value) -> char {
    match v {
        Value::V0 => '0',
        Value::V1 => '1',
        Value::X => 'x',
        Value::Z => 'z',
    }
}

/// Recursively walk the VCD scope tree and populate `out` with:
/// `IdCode → (full_hierarchical_name, scope_path, bit_width)`
fn collect_signals(
    items: &[ScopeItem],
    path: &str,
    out: &mut HashMap<IdCode, (String, String, u32)>,
) {
    for item in items {
        match item {
            ScopeItem::Scope(scope) => {
                let child_path = if path.is_empty() {
                    scope.identifier.clone()
                } else {
                    format!("{}.{}", path, scope.identifier)
                };
                collect_signals(&scope.items, &child_path, out);
            }
            ScopeItem::Var(var) => {
                let full_name = if path.is_empty() {
                    var.reference.clone()
                } else {
                    format!("{}.{}", path, var.reference)
                };
                out.insert(var.code, (full_name, path.to_string(), var.size));
            }
            _ => {}
        }
    }
}

// ── VCD parser ───────────────────────────────────────────────────────────────

pub fn parse_vcd(file_path: &str) -> Result<WaveformMeta> {
    let file =
        File::open(file_path).with_context(|| format!("Cannot open file: {}", file_path))?;
    let reader = BufReader::new(file);
    let mut parser = Parser::new(reader);

    let header = parser.parse_header().context("Failed to parse VCD header")?;

    let ts_factor_ns = match header.timescale {
        Some((magnitude, unit)) => timescale_to_ns(magnitude, unit),
        None => 1.0, // default: assume 1 ns
    };

    // Walk scope tree to collect signal metadata
    let mut id_to_info: HashMap<IdCode, (String, String, u32)> = HashMap::new();
    collect_signals(&header.items, "", &mut id_to_info);

    // Stream through simulation commands
    let mut tv_map: HashMap<IdCode, Vec<(f64, String)>> = HashMap::new();
    let mut current_time: u64 = 0;
    let mut end_time: u64 = 0;

    for cmd in parser {
        match cmd.context("Failed to parse VCD command")? {
            Command::Timestamp(t) => {
                current_time = t;
                if t > end_time {
                    end_time = t;
                }
            }
            Command::ChangeScalar(id, val) => {
                let t_ns = current_time as f64 * ts_factor_ns;
                tv_map
                    .entry(id)
                    .or_default()
                    .push((t_ns, value_to_char(&val).to_string()));
            }
            Command::ChangeVector(id, vals) => {
                let t_ns = current_time as f64 * ts_factor_ns;
                let val_str: String = vals.iter().map(|v| value_to_char(&v)).collect();
                tv_map.entry(id).or_default().push((t_ns, val_str));
            }
            Command::ChangeReal(id, val) => {
                let t_ns = current_time as f64 * ts_factor_ns;
                tv_map
                    .entry(id)
                    .or_default()
                    .push((t_ns, format!("{:.6}", val)));
            }
            _ => {}
        }
    }

    let duration_ns = end_time as f64 * ts_factor_ns;

    // Assemble final signals map
    let mut signals: HashMap<String, SignalTrace> = HashMap::new();
    for (id, (full_name, scope, width)) in id_to_info {
        let tv = tv_map.remove(&id).unwrap_or_default();
        signals.insert(
            full_name.clone(),
            SignalTrace {
                name: full_name,
                width,
                scope,
                tv,
            },
        );
    }

    Ok(WaveformMeta {
        format: "vcd".to_string(),
        file_path: file_path.to_string(),
        duration_ns,
        signals,
    })
}

// ── FST parser (via fst2vcd) ─────────────────────────────────────────────────

pub fn parse_fst(file_path: &str) -> Result<WaveformMeta> {
    let available = std::process::Command::new("which")
        .arg("fst2vcd")
        .output()
        .map(|o| o.status.success())
        .unwrap_or(false);

    if !available {
        bail!(
            "fst2vcd not found. Install GTKWave (e.g. `brew install gtkwave`) to enable FST parsing."
        );
    }

    let tmp_path = std::env::temp_dir().join(format!("waveform_mcp_{}.vcd", std::process::id()));
    let tmp_str = tmp_path.to_str().unwrap();

    let out = std::process::Command::new("fst2vcd")
        .args(["-o", tmp_str, file_path])
        .output()
        .context("Failed to run fst2vcd")?;

    if !out.status.success() {
        bail!(
            "fst2vcd conversion failed: {}",
            String::from_utf8_lossy(&out.stderr)
        );
    }

    let mut meta = parse_vcd(tmp_str)?;
    meta.format = "fst".to_string();
    meta.file_path = file_path.to_string();
    let _ = std::fs::remove_file(&tmp_path);

    Ok(meta)
}

// ── Auto-detect format ───────────────────────────────────────────────────────

pub fn parse_waveform(file_path: &str) -> Result<WaveformMeta> {
    let ext = std::path::Path::new(file_path)
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| e.to_lowercase());

    match ext.as_deref() {
        Some("vcd") => parse_vcd(file_path),
        Some("fst") => parse_fst(file_path),
        Some(e) => bail!(
            "Unsupported waveform format '.{}'. Supported: .vcd, .fst",
            e
        ),
        None => bail!("File has no extension; cannot determine waveform format"),
    }
}
